RestAPI 방식의 ToDo 관리 기능

BEOMKING opened the pull request:
## 요약

프론트 엔드와 백엔드가 RestAPI로 통신하며 ToDo 를 관리하는 기능입니다.

## 세부
Todo는 코드 내에서 Task로 사용되며 Id와 Title을 가집니다.

ToDo 관리 기능은 세부적으로 다음과 같습니다.

- ToDo 상세 조회
- ToDo 전체 조회
- ToDo 생성
- ToDo 수정
- ToDo 삭제

Commit:
Pull Request 생성 겸 리뷰어님에게 인사

BEOMKING committed:
해야 할 일이라는 의미의 Task 객체 생성

BEOMKING committed:
Task 저장소 생성

Task를 생성 수정 삭제하는 저장소를 만들었습니다.

BEOMKING committed:
TaskRepository add 메서드가 정확한 Task를 리턴하도록 변경, 테스트 코드
작성

get, add, remove 테스트 코드 작성
add Test 중 Task 대신 null을 리턴하는 것을 발견하고 add 메서드 변경

BEOMKING commented on app/src/test/java/com/codesoom/assignment/repository/TaskRepositoryTest.java:
```diff
@@ -0,0 +1,38 @@
+        assertEquals(new Task(0L, "BJP"), taskRepository.add("BJP"));
+    }
+
+    @Test
+    @DisplayName("입력 받은 taskId와 같은 id를 가진 Task가 존재하면, get 메서드가 해당 Task를 리턴한다.")
+    void getTestWhenFoundTaskId() {
+        taskRepository.add("BJP");
+        assertEquals(new Task(0L, "BJP"), taskRepository.get(0L));
+    }
+    @Test
+    @DisplayName("입력 받은 taskId와 같은 id를 가진 Task가 있을 때, remove 메서드가 해당 Task를 제거한다.")
+    void removeTestWhenInputIdEqualTaskId() {
+        taskRepository.add("BJP");
+        taskRepository.remove(0L);
+        assertNull(taskRepository.get(0L));
+    }
```
add 메서드를 구현하기 전까지 taskId와 같은 id를 가진 Task를 get, remove하는 테스트를 진행할 수 없었습니다.
이런 경우, add 메서드를 먼저 작성하고 테스트를 진행하는지 다른 방법이 있는지 궁금합니다.

johngrib replied:
방법은 무지 많습니다. 이럴 땐 트레이너가 있는 상황에서만 할 수 있는 것을 해보는 것도 좋겠죠. 제가 뭔가 방법을 알려드리면 그걸 해보는 것도 괜찮겠지만 아이디어를 몇가지 떠올려서 실험을 해보는 것입니다. 그러면 제가 코드를 읽고 코멘트를 드리겠죠? 트레이너가 있을 때 실험을 많이 해보세요. 제가 알려드려서 그걸 읽고 그대로 따라하는 건 당장은 좋겠지만 장기적으로는 아주 좋은 방법이 아닐 수도 있습니다.

하지만 이렇게 이야기하고 끝내면 좀 막막할 수 있을 테니 두 가지 방법을 짚고 생각의 재료로 삼아보는 것도 괜찮겠군요. 두 가지 방법을 생각해 봅시다.

제일 간단한 방법은 add 메소드 시그니처를 만들지만 method body는 비워놓고 null을 리턴하게 만들어서 동원하는 걸겁니다. 이러면 당연히 테스트는 실패하겠죠? 그러나 실패하고 있는 테스트를 보유하게 되니 이후로 목표는 테스트가 성공할 때까지 코드를 고치는 게 됩니다. TDD의 구색이 갖춰지게 되는 거죠.

다른 방법도 생각해 봅시다. 이건 위의 방법과 비슷하지만 구현을 하지 않았다는 사실을 Java 언어의 문법을 통해서 강하게 표현하는 방법입니다. 시그니처를 만들되 인터페이스로 만드는 거죠. 이러면 일부러 null 같은 걸 리턴하게 만들지 않아도 인터페이스를 테스트에 동원할 수 있습니다. 그 이후로 과제는 인터페이스의 구현체를 잘 만드는 것이 될 겁니다. 이렇게 하면 여러 구현체를 만들고 바꿔 끼워보면서 테스트를 돌려보게 될 수도 있겠죠.

BEOMKING replied:
215fbc1
TaskRepository 인터페이스를 만들고 TaskListRepository 를 구현하면서 테스트해봤습니다.
진행하면서 느낀 점은 굳이 `add 메서드 없이 모든 get 메서드 테스트가 통과되야 해` 하지 않아도 그때그때 필요한 테스트를 만들면서 구현해도 될 것 같아요.

BEOMKING committed:
Task와 관련한 로직을 처리하는 TaskService 생성

johngrib reviewed (commented):
PR에 바이너리 파일이 포함되어 있습니다. 이런 파일이 git commit에 추가되지 않도록 git add 를 사용할 때 주의하셔야 할 필요가 있어요. 특히 `git add .` 명령은 가급적 사용하지 않도록 습관을 들이시는 것이 좋습니다. 어떤 파일을 add 하는지 일일이 확인하고, 확인하면서 고민하는 시간을 가져 보세요. 몹시 귀찮고 시간낭비처럼 느껴질 수도 있어요. 하지만 필요한 과정입니다. 시간은 점점 줄어들거고 ,판단능력도 조금씩 향상될 거에요.

![image](https://user-images.githubusercontent.com/1855714/178261702-be8f4f90-5aad-4c5e-a397-1710f9d2fa82.png)

한편으로는 이렇게 추가한 파일을 git 히스토리에서 제외한 히스토리를 생성하는 연습을 해 볼 좋은 기회입니다. `git rm`이나 `git rebase -i` 에 대해 조사해 보세요.

johngrib reviewed (commented):
PR이 어떤 의미인지 생각해보는 것도 좋겠죠. Pull Request는 말하자면 관공서에 보내는 일종의 제안서와 비슷합니다. 내가 이런 작업을 했는데, 읽어보고 괜찮으면 승인도 하고 공식 작업물로 포함시켜 달라는 거죠. 그러니 제목도 작업의 내용을 잘 표현하는 것으로 수정해 보도록 합시다.

PR 타이틀을 대충 짓고 넘어가는 사람들이나 팀도 꽤 많은 편이긴 한데요, 우리는 더 좋은 것을 목표로 연습하고 있죠. 

![image](https://user-images.githubusercontent.com/1855714/178262230-9bd9045f-b176-4070-a91c-5be140238631.png)

PR 타이틀을 좋은 제목으로 바꾸고 나면 본문의 내용도 신경써서 작성해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/HelloController.java:
```diff
@@ -1,12 +1,13 @@
-    public String sayHello() {
-        return "Hello, world!";
+    @GetMapping("/{taskId}")
+    public String getTask(@PathVariable Long taskId) {
+        return "안녕하세요, 리뷰어님. 이번주도 잘 부탁드립니다.";
     }
```
저도 잘 부탁드립니다 :-)

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * 해야 할 일이라는 의미의 Task 클래스입니다.
+ */
+public class Task {
```
클래스를 설명할 때에는 `~클래스입니다.`라고 설명하지 않도록 합시다.

johngrib commented on app/src/test/java/com/codesoom/assignment/repository/TaskRepositoryTest.java:
```diff
@@ -0,0 +1,38 @@
+    @DisplayName("taskId를 받았지만 같은 id를 가진 Task가 존재하지 않을 때, get 메서드가 null을 리턴한다.")
+    void getTestWhenNotFoundTaskId() {
+        assertNull(taskRepository.get(0L));
+    }
```
설명과 테스트 코드의 내용이 서로 논리적으로 호응하도록 작업해주는 것이 좋습니다. 지금은 이 말이 어떤 의미인지 헷갈리실 수도 있어요. 하지만 열심히 생각해 보도록 합시다. 여기에서 더 나아간 내용은 조금 더 시간을 두고 같이 진행해 보죠.

johngrib reviewed (commented):
반갑습니다.

BEOMKING commented:
> PR에 바이너리 파일이 포함되어 있습니다. 이런 파일이 git commit에 추가되지 않도록 git add 를 사용할 때 주의하셔야 할 필요가 있어요. 특히 `git add .` 명령은 가급적 사용하지 않도록 습관을 들이시는 것이 좋습니다. 어떤 파일을 add 하는지 일일이 확인하고, 확인하면서 고민하는 시간을 가져 보세요. 몹시 귀찮고 시간낭비처럼 느껴질 수도 있어요. 하지만 필요한 과정입니다. 시간은 점점 줄어들거고 ,판단능력도 조금씩 향상될 거에요.
> 
> ![image](https://user-images.githubusercontent.com/1855714/178261702-be8f4f90-5aad-4c5e-a397-1710f9d2fa82.png)
> 
> 한편으로는 이렇게 추가한 파일을 git 히스토리에서 제외한 히스토리를 생성하는 연습을 해 볼 좋은 기회입니다. `git rm`이나 `git rebase -i` 에 대해 조사해 보세요.

Commit:
TaskRepository 인터페이스 생성, 기존 TaskRepository 자료구조에 종속된
명칭으로 변경, JavaDoc 존칭 통일

Commit:
입력 받은 id와 Task의 id를 비교해서 boolean을 던지는 메서드 추가

Commit:
TaskRepository, TaskMapRepository JavaDoc 변경, TaskListRepository 구현
및 테스트

Commit:
TaskRepository 인터페이스, 구현체에 getAll 메서드 시그니처 생성

Commit:
Map, List Repository getAll 메서드 구현, 테스트

Commit:
Task를 수정하는 메서드 생성

입력 받은 id와 같은 Task가 있을때, title을 변경시켜 리턴하는
메서드입니다.

BEOMKING committed:
컨트롤러에 입력 받은 taskId와 같은 id를 가진 Task를 리턴하는 getTask 메서드 생성

johngrib reviewed (commented):
커밋 메시지는 너무 길어지면 이렇게 잘리게 됩니다.
![image](https://user-images.githubusercontent.com/1855714/178487790-f9a4250a-96d8-4489-9e49-4a9a72a22ed2.png)

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,29 @@
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController("/tasks")
+public class TaskController {
+    private final TaskService taskService;
+
+    @Autowired
+    public TaskController(TaskService taskService) {
+        this.taskService = taskService;
+    }
+
+    /**
+     * Get 요청시 path에 taskId가 포함되어 있으면 서비스에서 taskId와 id가 같은 Task를 리턴한다.
+     *
+     * @param taskId 입력 받은 Long 타입의 taskId
+     * @return 입력 받은 taskId와 같은 id를 가진 Task 리턴
+     */
+    @GetMapping("/{taskId}")
+    public Task getTask(@PathVariable Long taskId) {
+        return taskService.getTask(taskId);
+    }
```
JavaDoc에서는 메소드의 입출력에 대해 설명하고, 애노테이션은 메타 프로그래밍을 통해 코드가 어떻게 수동적으로 작동하는지를 보여주죠. 그러므로 JavaDoc에서는 애노테이션에 대한 내용은 생략하는 것을 권해드리고 싶어요. 애노테이션은 스스로를 충분히 잘 설명하니까요.

BEOMKING replied:
애노테이션은 그 자체로 설명이 된다 라는 말씀이군요!

johngrib replied:
그렇습니다! 사실 이건 애노테이션 뿐 아니라 코드 명명과 코드 디자인에도 재귀적으로 적용할 수 있어야 합니다. 그래서 반대로 생각해볼 수도 있어요. 만약 충분히 설명적이지 않은 "무언가"를 발견했다면 그걸 언젠간(또는 지금 당장) 고쳐야겠다고 생각하는 거죠. "냄새를 맡았다"라고 표현하는 경우들 중의 하나이기도 합니다.

BEOMKING committed:
저장소가 비어있고 getAll 메서드로 조회할 때, 빈 배열을 리턴하는 테스트
추가

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,44 @@
+     *
+     * @param title 입력 받은 문자열 title
+     * @return 바뀐 Task 리턴
+     */
```
이렇게 주석에 `Task`처럼 타입을 명시해주면 클래스 이름과 주석 사이에 커플링이 생깁니다. 만약 `Task` 클래스 이름을 `PersonalTask`로 이름을 바꾼다고 상상해 봅시다. 그렇다면 주석 안쪽에 있는 `Task`도 전부 찾아서 일일이 바꿔줘야 하는 거죠. 지금은 파일이 몇 개 없어서 괜찮지만 만약 그런 파일이 꽤 많으면 무척 골치아플 것입니다. 이걸 통해서 객체지향 원칙과 DRY 원칙이 주석에도 통한다는 사실을 생각해볼 수 있을 거에요.

BEOMKING replied:
현재 테스트 코드에도 메서드, 타입명을 명시해주는 경우가 많은데 고민해봐야겠어요.

BEOMKING replied:
오늘 과제를 진행하며 주석들에는 Task에 `요소`, `작업`이라는 단어들을 사용해봤는데 저는 `작업`이라는 단어도 괜찮은 것 같아요.
`요소`는 컬렉션에서는 나쁘진 않았다고 생각했는데 다른 사람에게 명확한 의미를 주긴 힘든 단어라고 생각이 드네요.

johngrib replied:
적절한 타이밍에 좋은 고민을 하시게 된 것 같아요. 이런 고민은 많이 해보셔야 합니다. 우리가 하는 일의 대부분은 이름을 잘 짓는 것만으로도 효율이 부쩍부쩍 올라가기도 합니다. 그래서 심볼을 그대로 적을지, 주석에 추가로 정보를 남겨둘지, 뭔가 설명을 할 것인지, 원본 타입을 지시하는 심볼의 이름을 바꿀 것인지, 별도로 흔히 알려져서 이해받기도 쉬운 패턴으로 코드를 redesign할지 등을 늘 고민하게 되죠. 그런데 여기서 중요한 것은 "언제나 통용되는 절대적 규칙이란 것이 없다"는 것입니다. 그래서 매번 작성할 때마다 어떨 때에는 타입을 주석에 넣기도 하고, 어떨 때에는 "할일" 처럼 한국어로 넣는 것을 선택하기도 하죠. 모든 것은 주위의 코드, 나아가 환경, 즉 작고 큰 컨텍스트를 고려하며 선택해야 하는 것입니다. 이 컨텍스트도 재귀적으로 둘러봐야 하는데요, 좀 더 나아가서는 회사에서 함께 일하는 동료들도 포함해 생각할 수 있어야 하기 때문입니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
```
`if`를 작성할 때에는 "언제나" 중괄호를 생략하지 않도록 합시다. `if`를 한 줄로 작성하지 않는 것도 습관으로 익혀두도록 합시다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskListRepository.java:
```diff
@@ -0,0 +1,44 @@
+
+/**
+ * TaskRepository의 List 구현입니다.
+ */
```
어느 타입의 구현인지 설명하는 것보다 이 클래스의 역할과 책임에 대해 설명해 보세요.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskListRepository.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    public Task add(String title) {
+        Task task = new Task(id++, title);
```
`id++`가 멀티 스레드 안전한지에 대해 고민해 볼 필요가 있겠군요.

BEOMKING replied:
현재는 요청이 몰려올 경우 문제가 발생할 수 있습니다.
쓰레드 세이프한 자료구조로 변경했습니다.

BEOMKING commented:
> 커밋 메시지는 너무 길어지면 이렇게 잘리게 됩니다. ![image](https://user-images.githubusercontent.com/1855714/178487790-f9a4250a-96d8-4489-9e49-4a9a72a22ed2.png)

커밋의 요약을 짧게 적고 구체적으로 적을 것이 있다면 한 줄 띄우고 본문을 추가하는 방식을 연습해보겠습니다.

Commit:
전체 조회 메서드 테스트 코드 추가

빈 저장소가 주어졌을 때 TaskService의 getAll 메서드가 빈 배열을
리턴하는지 확인하는 메서드를 작성했습니다.
표시명으로 getAll 메서드명, 리턴 타입을 정확히 명시하는 것 대신 역할에 대한 설명을
작성하였습니다.
알아보기 쉬우면서 메서드명, 리턴타입 변경의 변화에 영향을 받지 않을지
긴가민가하네요.

Commit:
Request 경로 수정, 전체 조회 메서드 추가

Commit:
TaskRepository 조회 메서드 리턴 타입 변경

TaskRepository의 조회 메서드 리턴 타입을 Optional로 변경했습니다.
변경한 이유는 찾은 Task가 null일 경우에 대해 유저의 요청을 처리해 줘야
한다고 생각했기 때문입니다.
해당 인터페이스의 주석, 구현체의 주석을 변경하였습니다.
구현체들의 리턴 타입을 Optional로 변경하고 테스트 코드에 반영했습니다.

Commit:
주석 변경, List 구현체 remove 메서드 수정

get 메서드 수정에 따라 remove 메서드도 수정했습니다.

BEOMKING commented on app/src/main/java/com/codesoom/assignment/repository/TaskListRepository.java:
```diff
@@ -37,8 +37,13 @@ public Task add(String title) {
     @Override
     public void remove(Long taskId) {
-        taskList.remove(get(taskId));
+        taskList.remove(get(taskId).get());
```
get 메서드의 리턴 타입을 Optional<Task>로 바꿔서 remove 메서드를 수정해야 했습니다.

현재 코드는 taskId와 같은 Task가 존재하는지 여부를 Service에서 확인을 했다는 전제하에 있습니다. 
Service의 역할이라고 생각했기 때문입니다.

그런데 이 방법으로 remove 메서드 안에서 get으로 호출한 Optional에 바로 .get()을 사용하니 Optional의 의미가 없어지는 거 같습니다. 

현재 생각나는 방법은 

1. Repository의 remove 메서드에도 null을 체크하는 로직을 만든다.
(Service에서 검증했는데 Repository에도 필요한가? 라는 의문이 듭니다.)
2. 파라미터로 Service에서 찾은 Task를 받는 remove 메서드를 하나 더 생성한다. 
(입력으로 숫자 타입을 받는 메서드는 그대로 남아야 해서 기존 문제는 사라지지 않을 것 같습니다.
하지만 Repository에서 null을 체크하는 로직은 필요 없어집니다.)

이렇게 두 가지가 있습니다. 어떻게 접근하는 것이 좋을까요?

BEOMKING replied:
remove 메서드 자체에서 루프를 돌며 taskId를 찾아 삭제하면 위 문제들을 고민할 필요가 없겠군요..

johngrib replied:
Service와 Repository의 역할을 떠올려보세요. 설명해보는 것도 좋겠죠. 지금 고민하고 있는 것은 어쩌면 두 컴포넌트의 책임에 대한 고민일 수도 있습니다.

Commit:
getTask 서비스, 컨트롤러 수정, TaskNotFound 예외 생성

Optional 타입으로 반환하여 null 예외처리를 해주었습니다.

BEOMKING committed:
remove 메서드 구현 변경

루프를 돌면서 taskId와 같은 id를 가진 Task를 모두 삭제하는 방식으로
변경했습니다.

BEOMKING committed:
Post 요청시 createTask 기능 추가, Task에 Getter 추가

사용자의 요청시 Task를 생성하고 리턴 하는 기능을 구현했습니다.
Jackson Binding을 위해 Task에 Getter를 추가했습니다.

BEOMKING committed:
TaskService가 Repository 구현체에 의존하던 것 수정

BEOMKING committed:
Task 주석 수정, equals 포매팅 수정

요소라는 단어보다는 작업이 더 잘어울리는 것 같아 바꿔보았습니다.
equals 중괄호 포매팅을 적용했습니다.

BEOMKING committed:
Repository 구현체 주석 수정

BEOMKING committed:
자료구조에 동시성 적용

BEOMKING committed:
Task 변경 컨트롤러 메서드 생성

컨트롤러에 Task를 변경할 수 있는 기능을 구현했습니다.
변경할 데이터를 가지고 있는 ChangeTaksRequest 객체를 생성했습니다.

BEOMKING committed:
Task 삭제 기능 추가, 미사용 import 제거

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,80 @@
+        return taskService.getTask(taskId);
+    }
+
+    /**
+     * 저장소에 저장된 모든 요소들을 컬렉션 타입으로 리턴합니다.
+     *
+     * @return 저장소에 저장된 컬렉션 타입의 요소
+     */
+    @GetMapping("/all")
+    public List<Task> getAll() {
+        return taskService.getAll();
+    }
```
1. 저장소는 서비스 안쪽에 캡슐화되었으니 컨트롤러에서는 모르는 것이 바람직합니다. 심지어 주석에도 알려지 있지 않은 것이 바람직합니다. JavaDoc 주석은 대부분의 경우에 내부 구조를 알 필요 없는 사람이 읽는 글이라 생각하세요.
2. "저장소", "모든 요소", "컬렉션 타입"을 살펴보면 이 주석의 단어 대부분이 `getAll` 메소드의 역할을 고려하지 않고 Java 언어 코드를 강하게 의식한 것이라는 사실을 알 수 있습니다. 이 메소드가 사실은 그냥 "할 일 목록을 제공하는" 녀석이라는 것을 설명해 주는 것이 적절하겠죠

BEOMKING replied:
언어에 얽매이지 않으며 책임과 역할을 고려하여 작성하는 것이 정말 어렵군요! ㅎㅎ

johngrib replied:
맞아요 ㅎㅎㅎ 우리 직업이 어려운 이유 중의 하나인 것 같습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,80 @@
+        return taskService.createTask(title);
+    }
+
+    /**
+     * 입력 받은 변경 정보를 가지고 있는 객체를 받아 작업을 변경하고 변경한 작업을 리턴합니다.
+     *
+     * @param request 입력받은 변경 정보 request
+     * @return 변경된 작업 리턴
+     */
```
`Object`를 직접 다루는 라이브러리나 유틸리티 메소드를 만드는 경우에만 JavaDoc 주석에서 "객체"를 언급하도록 해보세요. 이 주석은 필요 이상으로 추상화되어 있습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/Task.java:
```diff
@@ -0,0 +1,70 @@
+        this.title = title;
+    }
+
+    /**
+     * Jackson Binding을 위한 메서드입니다.
+     *
+     * @return 요소의 id
+     */
```
`Task` 클래스의 주석을 읽는 사람이 여기에서 `Jackson`을 발견하게 될 겁니다. 이 메소드를 코딩할 때에는 Jackson을 염두에 두고 작업했을 수 있습니다. 그러나 JavaDoc에 남겨지는 것은 곤란합니다. Task는 Jackson보다 오래 지속되는 프로젝트의 논리적 구심점이고 수명이 더 긴 핵심 타입이기 때문이죠. SpringBoot의 구성은 굉장히 튼튼하게 여겨지지만 어떤 부속 라이브러리라도 교체될 수 있다는 전체하에 코딩하는 것이 바람직합니다. 특히 출력 라이브러리라면 좀 더 그렇겠죠. Task는 JSON 출력만을 위한 게 아니니까요.

BEOMKING replied:
현재 Task는 getter 메서드가 필요없지만 Jackson 때문에 강제로 만들었고, getter를 사용하지 않도록 명시하기 위해 해당 주석을 사용했습니다.

주석에서 라이브러리에 대한 내용을 명시하지 않는다면 다른 사람들에게 getter를 사용하지 못하도록 하기 위해서 어떤 방법을 사용할 수 있을까요?

johngrib replied:
Jackson을 사용하는 이유는 JSON으로 "출력"하기 위해서일 것입니다. 그렇다면 여기에서 고려해야 할 점이 하나 더 있습니다. 출력의 형태가 여러 가지가 될 수 있다는 거죠. JSON으로 출력하는 것 외에 XML로 출력하는 경우와, 종이 프린터로 출력하는 경우와, 이메일로 전송하기 위해 출력하는 경우, SMS로 출력해야 하는 경우를 모두 적용해야 하는 상황이 한 달마다 찾아온다고 가정해 봅시다. 그렇다면 여기에서 `Task`는 각 출력 타입의 원천이 되는 자료가 될 겁니다. 이때 사용할법한 방법이 여러가지가 있을텐데, Task를 "갖도록" (has a 관계가 되도록) Task를 중심으로 데코레이션 패턴이나 어댑터 패턴을 사용하는 방법이 그 중 대표적이라 할 수 있겠죠. 좀 전통적으로는 상속을 사용할 수도 있을 겁니다.

요약해 말하자면 지금은 JSON 하나만 생각하고 있기 때문에 Task의 책임이 Jackson과 밀접한 것처럼 보이는 것입니다. 그렇기 때문에 Task 클래스 안쪽에 Jackson을 강하게 의식한 주석이 남겨지는 거죠. 여기에서 떠올려볼 법한 중요한 격언이 있는데 "데이터는 코드보다 수명이 길다"라고 할 수 있을 것입니다. 만약 어떤 DB 테이블에 "이건 Java의 Task 클래스와 연결됩니다" 같은 주석을 남겨놨다고 합시다. 데이터는 코드보다 수명이 기니까, 이 데이터는 Java 코드보다 오래 살아남을 가능성이 높을 거에요. 나중에 Java가 아니라 JavaScript 코드로 이 데이터를 취급하게 될 수도 있겠죠. 이와 비슷한 문제인 것입니다.

그렇다면 이번엔 위에서 언급한 요구사항들이 실제로 있다고 작업해 보도록 할까요? JSON 출력과 XML 출력이 되도록 한 번 작업해보는 겁니다. 일종의 서브 과제라고 생각하고 작업을 해보세요. Task 가 JSON 으로 출력되도록 하는 것이 이번주 과제의 일부라면, XML 출력도 가능하도록 어떤 클래스 하나를 정의해 보세요.

BEOMKING replied:
Jackson의 xml 바인딩 라이브러리를 추가하였고 적용할 수 있는 User 클래스를 정의했습니다.

johngrib replied:
전 라이브러리를 추가하는 게 아니라 직접 작업해보는 저수준 작업을 연습해보시길 바랬는데 다른 방법으로 해결하셨군요. 물론 괜찮습니다 ㅎㅎ 이것도 해결 방법이죠.

저와 함께 주간 세션을 진행할 때 제가 이렇게 해보세요 저렇게 해보세요 라고 추가 학습을 권해 드린다면 기본적으로 라이브러리를 사용하는 것이 아니라 plain java로 구현해볼 것을 권하는 거라고 생각하고 접근해보시면 됩니다. 물론 그렇게 안 하셔도 괜찮습니다. 추가 학습이니까요. 추가 학습을 진행하시면 단계적으로 제가 계속 추가 과제를 내드리는데요, Java/Spring과는 별도의 과제라 할 수 있으며 어디까지나 선택이므로 중간에서 그만두고 코드숨 과제 자체로 돌아가는 것도 괜찮습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskListRepository.java:
```diff
@@ -0,0 +1,52 @@
+     * @return taskId와 같은 id의 요소가 존재하면 해당 요소 리턴, 존재하지 않으면 null 리턴
+     */
+    @Override
+    public Optional<Task> get(Long taskId) {
+        return taskList.stream()
+                .filter(t -> t.isSame(taskId))
+                .findFirst();
+    }
```
task를 리스트로 관리하면 이런 종류의 탐색을 피할 수 없을 겁니다. 여기에서 문제는 할 일이 3억 개 정도 등록됐을 때 제일 마지막에 있는 할일만 계속 조회하는 사람이 등장하는 상황이 있을 수 있다는 거죠. 그리고 할 일은 계속 추가되고, 오래된 할 일은 게으르거나 서비스에 로그인하는 것을 까먹은 사용자들에 의해 종종 방치될 겁니다.

BEOMKING replied:
맞습니다. 그래서 현재 Map 자료구조 구현체를 사용하고 있습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskMapRepository.java:
```diff
@@ -0,0 +1,61 @@
+    private Long id = 0L;
+    private final Map<Long, Task> taskMap = new ConcurrentHashMap<>();
+
+    /**
+     * 요청된 숫자 형식의 taskId와 같은 id를 가진 요소가 존재하면 요소를 리턴하고 아니면 null을 리턴합니다.
+     *
+     * @param taskId 요청된 숫자 형식의 taskId
+     * @return taskId와 같은 id를 가진 요소가 존재하면 요소 리턴, 존재하지 않으면 null 리턴
+     */
+    public Optional<Task> get(Long taskId) {
+        return Optional.ofNullable(taskMap.get(taskId));
+    }
```
주석과 리턴 타입에 대한 설명이 다릅니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskMapRepository.java:
```diff
@@ -0,0 +1,61 @@
+        return Optional.ofNullable(taskMap.get(taskId));
+    }
+
+    /**
+     * 현재 저장된 Task 저장소를 List 타입으로 변환하여 리턴한다.
+     *
+     * @return 변환된 List 타입의 Task 저장소
+     */
+    public List<Task> getAll() {
+        return new ArrayList<>(taskMap.values());
+    }
```
"저장소를 ~ 타입으로 변환하여"가 아쉽습니다. 메소드 안쪽에서 어떤 알고리즘으로 작동하는지에 대해 JavaDoc에서 설명하지 않는 것이 좋습니다. 알고리즘을 교체하게 되면 주석도 변경하게 되겠죠. 주석에는 최소한의 핵심적인 진실과 책임만 표현해 주도록 합시다.
만약 헷갈린다면 JavaDoc 주석도 일종의 프론트엔드라고 생각해 보시면 도움이 될 거에요. 어떤 것은 보여줘야 하지만 어떤 것은 굳이 보여줄 필요가 없죠. 우리가 작성하는 모든 코드 조각은 "적절한" 블랙박스가 될 수 있어야 한다고 생각하고 코드를 디자인해 봅시다.

Commit:
주석을 메서드에 역할을 고려하여 작성

BEOMKING committed:
메서드 시그니처와 맞지 않는 주석 수정, 언어에 얽매인 주석 수정

BEOMKING committed:
유효하지 않은 taskId 를 조회할 때, 예외 발생 테스트

BEOMKING committed:
존재하지 않는 작업 조회시 메시지를 담은 예외를 던집니다.

BEOMKING committed:
예외 메시지를 Response에 포함되도록 하는 설정 추가

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,88 @@
+        this.taskService = taskService;
+    }
+
+    /**
+     * 입력 받은 숫자 타입의 taskId와 같은 id를 가진 작업을 조회해 리턴한다.
+     *
+     * @param taskId 입력 받은 숫자 타입 taskId
+     * @return 입력 받은 taskId와 같은 id를 가진 작업 리턴, 존재하지 않으면 예외 메시지를 던짐
+     * @throws ResponseStatusException 찾는 작업이 존재하지 않으면 던집니다.
+     */
```
다음 JavaDoc 주석과 비교해 보세요. 아래는 제가 방금 작성한 것입니다. 어떤 점이 다른가요?

```suggestion
    /**
     * 주어진 식별자를 갖는 작업을 찾아 리턴한다.
     *
     * @param taskId 작업의 식별자
     * @return 작업
     * @throws ResponseStatusException 작업을 찾지 못한 경우
     */
```

BEOMKING replied:
주석을 보고 짧은 시간 안에 해당 메서드가 어떤 메서드인지 알 수 있었으며 파라미터, 객체의 변경에 닫혀있습니다.

BEOMKING committed:
Update app/src/main/java/com/codesoom/assignment/controllers/TaskController.java

조언해주신 간결함과 변경에 닫힌 주석 적용

Co-authored-by: JohnGrib <johngrib82@gmail.com>

BEOMKING committed:
전역 예외처리 로직 생성
app/src/main/java/com/codesoom/assignment/exception/ErrorResponse.java# Please enter the commit message for your changes. Lines starting

Commit:
TaskController 주석 추가

Commit:
Merge branch 'BEOMKING' of https://github.com/BEOMKING/spring-week2-assignment-1 into BEOMKING

Commit:
생성 삭제 HttpStatus 삽입

POST, DELETE시 HttpStatus를 201, 204를 리턴하게 하였습니다.
불필요한 상속과 프로덕트 코드를 반영하지 않던 테스트코드를
변경하였습니다.

Commit:
식별자를 가지고 있는 작가 생성

Commit:
유저를 생성하는 로직 생성

작가라는 의미의 클래스를 유저로 변경했습니다.
Controller, Service, Repository 구조로 구현했습니다.
유저를 생성하는 로직을 생성했습니다.
XML로 바인딩 해주는 라이브러리를 의존성에 넣었습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -15,6 +15,10 @@
 import java.util.List;
 
+/**
+ * REST를 관리하며 비즈니스 로직의 인터페이스 역할을 한다.
+ * 외부에서 사용할 수 있도록 작업과 관련한 기능을 노출할 책임을 가지고 있다.
```
REST를 관리한다기보다는 HTTP 프로토콜 요청 처리를 담당한다고 표현하는 것도 좋을 것 같습니다. REST의 정의에 대해 생각해 볼 기회도 되겠군요.

BEOMKING replied:
REST는 소프트웨어 아키텍처의 형식이기 때문에 REST를 관리한다는 표현은 옳지 않네요!
말씀해주신 것처럼 컨트롤러는 HTTP 요청을 처리하는 것이 올바른 표현이라고 생각합니다.

johngrib replied:
좋습니다. 한편으로 이 컨트롤러는 anti corruption layer의 역할도 하는 셈이 됩니다. HTTP라는 외부 개념이 도메인 레이어 안으로 침투하는 것을 컨트롤러가 막아주고 있는 거죠.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskMapRepository.java:
```diff
@@ -20,29 +20,29 @@ public class TaskMapRepository implements TaskRepository {
 
     /**
-     * 요청된 숫자 형식의 taskId와 같은 id를 가진 요소가 존재하면 요소를 리턴하고 아니면 null을 리턴합니다.
+     * 요청된 받은 숫자 타입 taskId와 같은 id를 가진 작업을 리턴합니다. 이 작업은 null일 수 있습니다.
```
리턴 타입이 Optional이므로 `null`일 수 있다는 것을 JavaDoc에서 굳이 표현해주지 않아도 됩니다. 보충 학습으로 `Optional`의 JavaDoc도 한 번 읽어보면 좋을 거에요.

BEOMKING replied:
Optional은 자체로 null이 아닌 값을 포함하거나 포함하지 않을 수 있는 컨테이너 객체라고 합니다.
그렇다면 Optional 자체로 null일 수도 있다는 것을 의미하므로 불필요한 주석이었네요!

johngrib replied:
좋습니다. 꼼꼼하게 읽어보신 것 같아 기쁩니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/domain/User.java:
```diff
@@ -0,0 +1,32 @@
+        this.name = name;
+    }
+
+    /**
+     * 출력을 위한 메서드입니다.
+     *
+     * @return 식별자
+     */
```
이 JavaDoc은 삭제해도 괜찮을 것 같습니다.

BEOMKING replied:
질문이 있습니다.
저는 get 메서드를 바인딩을 위해 사용했고 이 메서드를 다른 코드에서 호출되는 것은 원치 않습니다.
하지만 다른 사람과 협업을 하게 된다면 이 get 메서드가 있기 때문에 사용해도 되겠지 하는 생각으로 사용할 수도 있어보여요.
이런 경우 어떻게 저의 의도를 나타낼 수 있을까요?
스프링부트에서 Jackson을 사용하기 때문에 어쩔 수 없는 걸까요?

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/UserRepository.java:
```diff
@@ -0,0 +1,29 @@
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * 유저들의 저장소입니다.
+ * 유저에 대한 삽입, 조회, 갱신, 삭제을 수행할 책임을 가지고 있습니다.
+ */
```
좋습니다. 한편으로는 Repository라는 단어 자체에 대한 학습도 필요할 수 있습니다. 여유가 있다면 에릭 에반스의 도메인 주도 설계 책에서 Repository가 등장하는 곳들을 찾아 읽어보는 것도 괜찮을 겁니다.

johngrib commented:
벌써 토요일이군요. 주간 회고를 작성해 공유해 주세요.

Commit:
코드숨 API 통과를 위한 수정

컨트롤로 API 경로 수정, remove 메서드 검증 추가

Commit:
RequestBody 입력 잘못 받던 것 수정

BEOMKING committed:
CrossOrigin 설정

BEOMKING committed:
PatchMapping 추가

BEOMKING committed:
도움 받아 PUT, PATCH 정보 동시에 처리

BEOMKING committed:
컨트롤러의 역할에 관한 맞지 않는 주석 수정

BEOMKING committed:
get 메서드의 주석 수정

Optional은 자체로 null을 가질 수 있다는 의미를 가지고 있으므로 주석을 수정합니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/controllers/TaskController.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    /**
+     * 입력 받은 숫자 형식의 taskId와 같은 taskId를 가진 작업을 제거합니다.
```
입력 파라미터 타입이 메소드 시그니처에 명시되어 있으니 타입이 어떤 형식이어야 하는지는 JavaDoc에서 굳이 표현해주지 않아도 괜찮습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskListRepository.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    /**
+     * 숫자 타입의 taskId가 들어오고 해당 taskId와 같은 task가 존재할 때, 해당 Task를 찾아 삭제한다.
```
제 취향대로라면 이렇게 할 것 같습니다. (취향이니 참고만 하셔도 괜찮아요)

```suggestion
     * 작업을 삭제합니다.
```

johngrib commented on app/src/main/java/com/codesoom/assignment/repository/TaskMapRepository.java:
```diff
@@ -0,0 +1,61 @@
+    private Long id = 0L;
+    private final Map<Long, Task> taskMap = new ConcurrentHashMap<>();
+
+    /**
+     * 같은 식별자를 가진 작업을 리턴합니다.
+     *
+     * @param taskId 식별자
+     * @return 작업 리턴
+     */
```
`TaskMapRepository`는 `TaskRepository`의 JavaDoc을 상속합니다. 그래서 겹치는 JavaDoc은 적절히 삭제해 주셔도 좋습니다.

BEOMKING replied:
JavaDoc 또한 상속이 되는군요. 중복된 주석은 삭제했습니다.

johngrib commented on app/src/main/java/com/codesoom/assignment/service/UserService.java:
```diff
@@ -0,0 +1,30 @@
+public class UserService {
+    private final UserRepository userRepository;
+
+    @Autowired
```
`@Autowired`의 사용은 권장되지 않습니다. 이건 스프링 공식 문서의 어딘가에 나와있는데요, 한번 찾아서 읽어보도록 해보세요. 찾는 과정도, 찾아서 읽는 과정도 좋은 공부가 될 것입니다. 그리고 문서에서 제안하고 있는 대안을 적용해 보도록 해보세요.

johngrib commented on app/src/test/java/com/codesoom/assignment/repository/TaskRepositoryTest.java:
```diff
@@ -0,0 +1,64 @@
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class TaskRepositoryTest {
```
테스트 코드를 좀 더 구조화할 필요가 있어 보입니다. 이번주는 이제 끝났으니 다음번에 같이 진행해보도록 해요.

johngrib reviewed (commented):
한 주간 고생 많으셨습니다.

BEOMKING committed:
Update app/src/main/java/com/codesoom/assignment/repository/TaskListRepository.java

Co-authored-by: JohnGrib <johngrib82@gmail.com>

BEOMKING committed:
주석 수정

BEOMKING committed:
Merge branch 'BEOMKING' of https://github.com/BEOMKING/spring-week2-assignment-1 into BEOMKING

BEOMKING committed:
주석 수정

BEOMKING committed:
주석 수정

주석을 간결하게 수정하였습니다.
JavaDoc은 상속되기 때문에 TaskRepository를 구현하는 클래스들의 중복된 주석은 제거하였습니다.
